### PR TITLE
Fix issue #789. Unknown keyword argument raises an error

### LINKED ISSUE #789: Warn if invalid keyword arguments are passed to a function
The following meson.build file triggers no errors, either with Meson 0.31.0 or with current 'master' commit a2e7ebc575a47bfb3:

```
project('testcase', 'c', foo: 'nonsense')
declare_dependency(grr: 1, bark: 2, woof: 3)
```

This is causing issues for me because I'll do something dumb like this (in a real project) without realising:

```
 widgets_dep = declare_dependency(
  link_with: librbwidgets,
  declare_dependencies: [glib, gstreamer, gtk]
)
```

and then I'll run a build and see that something, somewhere went wrong in the long chain of dependencies within this project and the Gtk compile flags are being lost ... but with no hint as to where the error is, it often takes a while debugging to figure that out.

https://github.com/mesonbuild/meson/issues/3 suggests that Meson would warn on invalid keyword arguments, but it doesn't.

### PR DESCRIPTION
**Pull request by ryobg**
Hello all, 
please review my shyly try to fix #789.
I'm not very sure what should be the policy when  the interpreter sees invalid keyword argument, but so far it seems to be warning or error. I chose the error path. Let me know if I can improve.
Thank you!

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -1413,24 +1413,26 @@ def func_files(self, node, args, kwargs):
...
-        deps = kwargs.get('dependencies', [])
+        deps = kwargs.pop('dependencies', [])
         if not isinstance(deps, list):
             deps = [deps]
-        compile_args = mesonlib.stringlistify(kwargs.get('compile_args', []))
-        link_args = mesonlib.stringlistify(kwargs.get('link_args', []))
+        compile_args = mesonlib.stringlistify(kwargs.pop('compile_args', []))
+        link_args = mesonlib.stringlistify(kwargs.pop('link_args', []))
+        if len(kwargs) > 0:
+            raise InvalidArguments('Unknown keyword arguments')
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: also, you should print which ones.

**Comment by ryobg** [comment]
What is with the comment?
As for the print I didn't want to overdo. I saw also #151 and thought - maybe there is a need of more generic solution?

**Comment by jpakkane** [comment]
Thanks for this. However I'd really like to have a more concentrated solution probably based on decorators. Currently each function does kinda its own thing for validation which is not really nice. Preferably we should be able to do something like this:

    @permittedKwargs('foo', 'bar', 'baz')
    def func_something(self, ...)

This ties into our goals of trying to generate documentation directly from the code rather than maintaining a separate wiki.

Functions should also treat their arguments read only whenever possible, so doing `pop`s on them is a bit suspect.

**Comment by ryobg** [comment]
I don't mind doing that. Minor cons is that you can run into different kwargs in the decorator and the function, but that looks easily manageable. I would add that the check can run as pre-condition saving up from any effects done in the actual function.

**Comment by ryobg** [comment]
What you would say? This is more proof of concept commit. I saw also few `pop` here and there, but that's off topic now. If fine I can go and use that everywhere in that file where kwargs are used. Btw, I think it is good idea to print not only the errors, but also the list of allowed arguments. Can be helpful. Issue is that the error message can become harder to read.

**Comment by QuLogic** [comment]
> However I'd really like to have a more concentrated solution probably based on decorators.

You could go the pytest route with introspection of args/kwargs, and if only using 3.5, could use type hints... Too bad that's not the minimum, though.

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
This has been superseded by #1988. Thanks.

**Pull request closed without merging**